File: UVa_ACM/100 - Volume C/UVa 10066.cpp
#include <iostream>
using namespace std;

static int n1[101],n2[101],a[101][101],i,j,t;

int main() {
	for (t=1;cin>>n1[0]>>n2[0]&&n1[0]&&n2[0];t++) {
		a[0][0]=0;
		for (i=1;i<=n1[0];i++) {cin>>n1[i];a[i][0]=0;}
		for (j=1;j<=n2[0];j++) {cin>>n2[j];a[0][j]=0;}
		for (i=1;i<=n1[0];i++) {
			for (j=1;j<=n2[0];j++) {
				if (n1[i]==n2[j]) a[i][j]=a[i-1][j-1]+1;
				else a[i][j]=max(a[i-1][j],a[i][j-1]);
			}
		}
		cout<<"Twin Towers #"<<t<<endl<<"Number of Tiles : "<<a[n1[0]][n2[0]]<<endl<<endl;
	}
}

File: since2020/UVa/UVa 10066.py
'''
	@judge UVa
	@id 10066
	@name The Twin Towers

	@tag Longest Common Subsequence
'''
from sys import stdin

dp = [ [ 0 ] * 101 for _ in range(101) ]

t = 1
for x, a, b in zip(*[ iter(stdin) ] * 3):
	n, m = map(int, x.split())
	xs = list(map(int, a.split()))
	ys = list(map(int, b.split()))
	for x in range(1, n + 1):
		for y in range(1, m + 1):
			if xs[x - 1] == ys[y - 1]:
				dp[x][y] = dp[x - 1][y - 1] + 1
			else:
				dp[x][y] = max(dp[x - 1][y], dp[x][y - 1])
	print('Twin Towers #{}\nNumber of Tiles : {}\n'.format(t, dp[n][m]))
	t += 1